### ISSUE TITLE
✅ 🧩 🐞 Bug Report: 3rd-party widgets using deprecated code don't work after updating Elementor to version 3.16.0 - [ED-12176]

### ISSUE THREAD

**Issue by jontng** [elementor/elementor#23745]
### Prerequisites

- [X] I have searched for similar issues in both open and closed tickets and cannot find a duplicate.
- [X] The issue still exists against the latest stable version of Elementor.

### Description

If you upgrade to Elementor version 3.16, custom widgets are no longer loaded.
If you go back to version 3.15.3, custom widgets are loaded again.

Elementor Pro 3.16 does not affect the problem.

### Steps to reproduce

1. upgrade to version 3.16
2. downgrade to version 3.15.3 (solution)

### Isolating the problem

- [ ] This bug happens with only Elementor plugin active (and Elementor Pro).
- [ ] This bug happens with a Blank WordPress theme active ([Hello theme](https://wordpress.org/themes/hello-elementor/)).
- [x] I can reproduce this bug consistently following the steps above.

### System Info

<details>
<summary>Click to reveal</summary>

````txt



== Server Environment ==
	Operating System: Darwin
	Software: Apache/2.4.43 (Unix)
	MySQL version: MySQL Community Server (GPL) v5.7.28
	PHP Version: 8.0.0
	PHP Memory Limit: 256M
	PHP Max Input Vars: 4000
	PHP Max Post Size: 1000M
	GD Installed: Yes
	ZIP Installed: Yes
	Write Permissions: All right
	Elementor Library: Connected

== WordPress Environment ==
	Version: 6.3.1
	Site URL: http://dev.dinglegymnasiet.se
	Home URL: http://dev.dinglegymnasiet.se
	WP Multisite: No
	Max Upload Size: 300 MB
	Memory limit: 40M
	Max Memory limit: 256M
	Permalink Structure: /%postname%/
	Language: sv-SE
	Timezone: Europe/Stockholm
	Debug Mode: Inactive

== Theme ==
	Name: Naturbruk Väst
	Version: 2023
	Author: The Bozzanova Agency Team
	Child Theme: No

== User ==
	Role: administrator
	WP Profile lang: sv_SE
	User Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36

== Active Plugins ==
	Advanced Custom Fields PRO
		Version: 6.2.1.1
		Author: WP Engine

	Elementor
		Version: 3.16.0
		Author: Elementor.com

	Elementor Pro
		Version: 3.15.1
		Author: Elementor.com

	Loco Translate
		Version: 2.6.6
		Author: Tim Whitlock

	Query Monitor
		Version: 3.13.1
		Author: John Blackbourn

	Safe SVG
		Version: 2.2.0
		Author: 10up

	Yoast Duplicate Post
		Version: 4.5
		Author: Enrico Battocchi & Team Yoast


== Elementanvändning ==
	
	archive : 1
		container : 2
		loop-grid : 1
		theme-archive-title : 2
	footer : 1
		container : 1
	wp-page : 5
		accordion : 1
		button : 2
		container : 54
		heading : 18
		image : 3
		text-editor : 12
		video : 1
	loop-item : 1
		container : 1
		heading : 1
		theme-post-featured-image : 1
		theme-post-title : 1


== Inställningar ==
	
	cpt_support: post, page
	allow_tracking: yes
	unfiltered_files_upload: 1
	meta_generator_tag: 1


== Funktioner ==
	Custom Fonts: 2
	Custom Icons: 0

== Integrationer ==
	


== Elementor-experiment ==
	Optimized DOM Output: Aktiv
	Förbättrad inläsning av filer: Aktiv
	Förbättrad CSS-laddning: Aktiv
	Inline Font Icons: Aktiv
	Ytterligare anpassade brytpunkter: Aktiv
	admin_menu_rearrangement: Inaktiv som standard
	Flexbox Container: Aktiv
	Uppgradera Swiper-bibliotek: Inaktiv som standard
	Grid Container: Inaktiv som standard
	Default to New Theme Builder: Inaktiv
	Editor Top Bar: Inaktiv som standard
	Landningssidor: Inaktiv
	Nested Elements: Aktiv
	Uppskjuten inläsning av bakgrundsbilder: Aktiv
	Global stilguide: Inaktiv som standard
	Sidövergångar: Inaktiv
	Notes: Inaktiv
	Loop: Aktiv
	Formulärinlämningar: Inaktiv
	Scroll Snap: Inaktiv
	Meny: Inaktiv som standard
	Taxonomy Filter: Inaktiv som standard



````

</details>

**Comment by perks98** [comment]
I've noticed some third-party plugins that were using 'register_widget_type' or 'widgets_registered' having widgets disappear in 3.16, in your custom widgets are you using those functions or any other depreciated Elementor functions? Just generally wondering myself

**Comment by APEFOOD** [comment]
> I've noticed some third-party plugins that were using 'register_widget_type' or 'widgets_registered' having widgets disappear in 3.16, in your custom widgets are you using those functions or any other depreciated Elementor functions? Just generally wondering myself

This is exactly it. I'm facing the same issue as @jontng and it seems to be because of deprecated functions. Would greatly appreciate to pointed to a change log resource that will show me all the updates I need to make to my code. This has not been a fun day at all. Deprecation or not, changing the API without a major version bump is not cool.

**Comment by jontng** [comment]
> I've noticed some third-party plugins that were using 'register_widget_type' or 'widgets_registered' having widgets disappear in 3.16, in your custom widgets are you using those functions or any other depreciated Elementor functions? Just generally wondering myself

Yes, it turned out to be the new action hook that was the problem.
The documentation should be a little more specific about these changes. Haven't seen any warnings about this change.

add_action( 'elementor/widgets/register', [ $this, 'init_widgets' ] ); // new hook
add_action( 'elementor/widgets/widgets_registered', [ $this, 'init_widgets' ] ); // old hook

**Comment by APEFOOD** [comment]
> > I've noticed some third-party plugins that were using 'register_widget_type' or 'widgets_registered' having widgets disappear in 3.16, in your custom widgets are you using those functions or any other depreciated Elementor functions? Just generally wondering myself
> 
> Yes, it turned out to be the new action hook that was the problem. The documentation should be a little more specific about these changes. Haven't seen any warnings about this change.
> 
> add_action( 'elementor/widgets/register', [ $this, 'init_widgets' ] ); // new hook add_action( 'elementor/widgets/widgets_registered', [ $this, 'init_widgets' ] ); // old hook

For anyone else who could be facing this problem, also make sure to replace all calls to:
> register_widget_type( new Your_Custom_Widget());  

with just: 
> register( new Your_Custom_Widget());

**Comment by soyket** [comment]
The hook was dedicated a long time ago (on 3.5 ) + the console was showing warnings about this update all the times if old hook were used so far and the guide also suggests using the new one - 
https://developers.elementor.com/v3-5-planned-deprecations/

https://developers.elementor.com/docs/managers/registering-widgets/

So not sure why developers of 3rd party didn't update them in the end.

**Comment by rami-elementor** [comment]
Elementor developer docs have code examples how to upgrade deprecated code:

https://developers.elementor.com/docs/deprecations/advanced-example/

**Comment by perks98** [comment]
For me, the 3.16.2 core update fixed my issue around this from what I can tell so far.

**Comment by APEFOOD** [comment]
> Elementor developer docs have code examples how to upgrade deprecated code:
> 
> https://developers.elementor.com/docs/deprecations/advanced-example/

Thank you @rami-elementor 🙏🏻

**Comment by nicholaszein** [comment]
Hello everyone,

### ✅ As you may have already seen, this issue was solved with the release of Elementor Core v3.16.2.

To clarify what was the issue, we implemented a solution to better handle deprecation notices, make them more human-readable, and easier to identify where they were used.

However, due to how some 3rd-party add-ons use the deprecated code, it was inadvertently causing an issue with the new implementation. We reverted this back and will try to work on a different way to handle deprecation notices.

### Disclosing Deprecations

It's important to note that the code used in question that was affecting 3rd-party widgets was deprecated with the release of **[Elementor Core version 3.5](https://wordpress.org/plugins/elementor/#developers:~:text=WordPress%20Dashboard%20screens-,3.5.0%20%E2%80%93%202021%2D12%2D14,-New%3A%20Introducing%20CSS)**. This version was released on December, 14th, 2021. **<ins>Almost 2 years, and 11 versions ago</ins>**.

At the time, we released a **[blog post outlining the deprecation](https://developers.elementor.com/v3-5-planned-deprecations/)**. We also have **[extensive documentation](https://developers.elementor.com/docs/getting-started/)**, and a public document outlining our **[Deprecation Process](https://developers.elementor.com/docs/deprecations/)**.

### Maintaining Compatibility

If you are a 3rd-party developer or if you have created any sort of integration, extension, or widget for Elementor, **<ins>it's imperative that you keep your code up-to-date to maintain full compatibility with the latest versions of Elementor<;ins>**, and to avoid issues like this.

If you use a 3rd-party add-on, it's important to keep an eye on the [Compatibility Tag](https://developers.elementor.com/compatibility-tag/) of ELementor to check if your add-on is compatible with the available version. If it's not, or if your add-on does not declare compatibility, it's recommended that you reach out to the developer and ask them to confirm the compatibility and include the compatibility tag.

### Contributing and Staying Informed

Elementor has a **Partnership Program** and tries its best to inform partners about changes that might affect their integration. But we can't control how 3rd-party developers handle their software.

We also run a **Public Beta** program for at least 3 weeks, and have a dedicated **[Developer Edition](https://wordpress.org/plugins/elementor-beta/)** plugin that developers can use to test their plugins against the newest changes in Elementor, and ensure that their plugins are compatible. This gives everyone plenty of time to make any necessary changes.

If you are an add-on developer, you can also [Subscribe to Product Updates](https://elemn.to/devloop), or you can subscribe to updates from our Developers Blog by enabling the **`Beta Tester`** option in the WordPress **`Dashboard`** > **`Elementor`** > **`Tools`** > **`Version Control`** page. When you enable this option, Elementor will ask if you want to subscribe to updates.

### We need your help!

We want our users to have the best experience they can with Elementor. We urge all developers to strive to achieve the same goal by maintaining their integrations compatible with the latest version of Elementor and Elementor Pro.

We hope that issues like this one do not repeat again.

### 🙏 Thank you all who reported this issue and helped us find a solution for it.

Kind regards.
